spring-integration-core/src/test/java/org/springframework/integration/dispatcher/PollingTransactionTests.java

```diff
@@ -196,7 +200,53 @@ public void propagationMandatory() throws Throwable {
 		Message<?> errorMessage = errorChannel.receive(3000);
 		assertNotNull(errorMessage);
 		Object payload = errorMessage.getPayload();
-		assertEquals(IllegalTransactionStateException.class, payload.getClass());
+		assertEquals(MessagingException.class, payload.getClass());
+		MessagingException messagingException = (MessagingException) payload;
+		assertEquals(IllegalTransactionStateException.class, messagingException.getCause().getClass());
+		assertNull(output.receive(0));
+		assertEquals(0, txManager.getCommitCount());
+		context.close();
+	}
+
+	@Test
+	public void commitFalilureTest() throws Throwable {
+		ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext(
+				"transactionFailureTests.xml", this.getClass());
+		TestTransactionManager txManager = (TestTransactionManager) context.getBean("txManagerBad");
+		PollableChannel input = (PollableChannel) context.getBean("inputTxFailure");
+		PollableChannel output = (PollableChannel) context.getBean("output");
+		PollableChannel errorChannel = (PollableChannel) context.getBean("errorChannel");
+		assertEquals(0, txManager.getCommitCount());
+		input.send(new GenericMessage<String>("commitFalilureTest"));
+		Message<?> errorMessage = errorChannel.receive(3000);
+		assertNotNull(errorMessage);
+		Object payload = errorMessage.getPayload();
+		assertEquals(MessagingException.class, payload.getClass());
+		MessagingException messagingException = (MessagingException) payload;
+		assertEquals(IllegalTransactionStateException.class, messagingException.getCause().getClass());
+		assertNotNull(messagingException.getFailedMessage());
+		assertNotNull(output.receive(0));
+		assertEquals(0, txManager.getCommitCount());
+		context.close();
+	}
+
+	@Test
+	public void handlerFalilureTest() throws Throwable {
+		ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext(
+				"transactionFailureTests.xml", this.getClass());
+		TestTransactionManager txManager = (TestTransactionManager) context.getBean("txManagerGood");
+		PollableChannel input = (PollableChannel) context.getBean("inputHandlerFailure");
+		PollableChannel output = (PollableChannel) context.getBean("output");
+		PollableChannel errorChannel = (PollableChannel) context.getBean("errorChannel");
+		assertEquals(0, txManager.getCommitCount());
+		input.send(new GenericMessage<String>("commitFalilureTest"));
+		Message<?> errorMessage = errorChannel.receive(3000);
```

Comment:
Usually our CI machine is pretty slow, so 3 secs might not be enough to wait for the answer from another thread.
Right, I see that it is a practice of this test class, but let's move on to the more robust solution.
So, 10 sec is our typical time.